### DISCUSSION TITLE
Repeated SAF prompts with self-built nightlies

### DISCUSSION THREAD

**Discussion by ztNFny** [cgeo/cgeo#13865] [Q&A]
Hi devs,
I only use c:geo build built myself and install them using "adb install cgeo-nightly.apk". After each install I get again c:geos file migration dialog which prompts a SAF dialog.
Obviously each build is signed with the same key.

Is this expected behaviour or am I doing something wrong with the builds / is there something wrong with the folder detection?

**Comment by eddiemuc** [comment]
I can't outrule the possibility of a bug of course. But normal behavior should be that c:geo only actively asks if it doesn't have folder permissions yet

**Reply by ztNFny** [reply]
Should it lose it on "adb install"? I assume not... Any pointers where to start debugging?

**Reply by Lineflyer** [reply]
IMHO you should use `adb install -r` to keep all preferences if the app is already installed.

**Reply by ztNFny** [reply]
> IMHO you should use `adb install -r` to keep all preferences if the app is already installed.

Not required, that's the default. But also installing with it or interactively, same behivour.

**Comment by eddiemuc** [comment]
* Can you reproduce in an emulator? (I assume that Android Studio also uses adb install in the background when installing a dev image on an emulator?)
* Can you ignore the migration when this happens (do not choose any SAF folder), then go to system information? -> there in the path section you should find persisted uri permissions (near the end of the section) -> would be interesting what can be seen there

**Reply by ztNFny** [reply]
"System information" I assume you mean the logfile?
After update and not re-allowing the folder it has:

- Public Folders: #12
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:false, files:0, dirs:0, totalFileSize:0 B, free space: -1 B, files on device: -1)

But no files from disk are being accessed (e.g. my map is missing)

**Reply by eddiemuc** [reply]
> "System information" I assume you mean the logfile? 

No, I actually meant the system information under "About c:geo"., and there the "persisted uri permissions" (but you ccan also just copy/paste the whole thing of course). In my emulator (which is set up very typical with one base folder containing everything of c:geo and one special fodler for maps) this looks like this:

![image](https://user-images.githubusercontent.com/6909759/212484370-b0a5e843-1a0e-458e-9042-24ddd213065b.png)

**Reply by ztNFny** [reply]
Ah, didn't know we had that dialog ;-) It's just displaying the logfile though...

- Persisted Uri Permissions: #6
  - content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%[2Fmenion.android.locus.pro](http://2fmenion.android.locus.pro/)%2FmapsVector (Sep 16, 2022, 22:58):RW
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo%2Fmaps%2F_themes (Dec 13, 2021, 21:34):RW
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (Jan 14, 17:30):RW
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo%2Fgpx (Dec 18, 2021, 19:08):RW
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo%2Frouting%2Fsegments4 (Dec 2, 2021, 19:49):RW
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo%2Fmaps%2F_hgt (Sep 16, 2022, 23:11):RW

The "primary%3Acgeo" is still listed there, even when it's not available at that moment. After granting it'll look the same but the timestamp will be updated.

**Reply by ztNFny** [reply]
Interestingly all other storage locations don't lose their permission, only the root folder does.

**Reply by eddiemuc** [reply]
The permissions all look ok to me. Very strange. Could be a c:geo bug. But I don't know how to reproduce

**Comment by ztNFny** [comment]
I checked adb logcat on installation and start. On installation nothing, on start there's this, but I think it's too late already at that point

01-14 18:53:40.832 20769 20769 E ActivityThread: Failed to find provider info for com.android.externalstorage.documents
01-14 18:53:40.833 20769 20769 W cgeo    : [main] ContentStorage: Folder 'BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]](content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo)' could not be accessed
01-14 18:53:40.833 20769 20769 D CompatibilityChangeReporter: Compat change id reported: 147798919; UID 10986; state: ENABLED

**Reply by eddiemuc** [reply]
The only thing that comes to my mind is that maybe Android needs some time after c:geo startup to grant permission to access a folder and that c:geo tries too soon to do it. But that would be quite stupid. Also, then access should work later in the app. 

Do you get more info when starting with debug log level?

**Reply by ztNFny** [reply]
Just more info in the same line

01-14 19:25:09.660 22519 22519 E ActivityThread: Failed to find provider info for com.android.externalstorage.documents
01-14 19:25:09.660 22519 22519 W cgeo    : [main] ContentStorage: Folder 'BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]](content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo)' could not be accessed {ContentStorage.reportProblem:641/ContentStorage.reportProblem:629/ContentStorage.getFolder:592/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28/Activity.performCreate:8129/Activity.performCreate:8109}

**Comment by eddiemuc** [comment]
Can you try to put this file in your logfile folder? It will configure an extremely fine-grained logging (if c:geo can read it...):

https://github.com/cgeo/cgeo/files/5213102/log-properties.txt

**Reply by ztNFny** [reply]
01-14 19:47:28.771 20325 20325 I cgeo    : [main] [CtxLog]ContentStorage.init:START
01-14 19:47:28.773 20325 20325 I cgeo    : [main] [CtxLog]ContentStorage.init:END (2ms)
01-14 19:47:28.776 20325 20355 D CompatibilityChangeReporter: Compat change id reported: 160794467; UID 10986; state: ENABLED
01-14 19:47:28.782 20325 20325 I cgeo    : [Log] Logging set: minLevel=WARN, minAddCallerInfo=NONE, addCallerInfoMaxDepth=8, throwOnError=false
01-14 19:47:28.784 20325 20325 E ActivityThread: Failed to find provider info for com.android.externalstorage.documents
01-14 19:47:28.786 20325 20325 I cgeo    : [Log] Logging config file 'log-properties.txt'found at /data/user/0/cgeo.geocaching/files/public/LOGFILES[FILE#2:p-file:///data/user/0/cgeo.geocaching/files::/public/LOGFILES], try to apply
01-14 19:47:28.787 20325 20325 I cgeo    : [Log] Logging set: minLevel=VERBOSE, minAddCallerInfo=VERBOSE, addCallerInfoMaxDepth=10, throwOnError=false
01-14 19:47:28.787 20325 20325 I cgeo    : [main] ---------------- CGeoApplication: startup ------------- {CgeoApplication.onCreate:51/Instrumentation.callApplicationOnCreate:1226/ActivityThread.handleBindApplication:7046/ActivityThread.access$1800:254/ActivityThread$H.handleMessage:2184/Handler.dispatchMessage:106/Looper.loopOnce:233/Looper.loop:344/ActivityThread.main:8212/Method.invoke:-2}
01-14 19:47:28.787 20325 20325 I cgeo    : [main] [CtxLog]CGeoApplication.onCreate:START {CgeoApplication.onCreate:52/Instrumentation.callApplicationOnCreate:1226/ActivityThread.handleBindApplication:7046/ActivityThread.access$1800:254/ActivityThread$H.handleMessage:2184/Handler.dispatchMessage:106/Looper.loopOnce:233/Looper.loop:344/ActivityThread.main:8212/Method.invoke:-2}
01-14 19:47:28.791 20325 20325 I cgeo    : [main] [CtxLog]DataStore.init:START {DataStore.initAndCheck:827/DataStore.init:812/DataStore.getLists:4058/CgeoApplication.onCreate:69/Instrumentation.callApplicationOnCreate:1226/ActivityThread.handleBindApplication:7046/ActivityThread.access$1800:254/ActivityThread$H.handleMessage:2184/Handler.dispatchMessage:106/Looper.loopOnce:233}
01-14 19:47:28.793 20325 20325 I cgeo    : [main] [DB] Current Database Version: 99 {DataStore$DbHelper.onOpen:1719/SQLiteOpenHelper.getDatabaseLocked:427/SQLiteOpenHelper.getWritableDatabase:316/DataStore.initAndCheck:834/DataStore.init:812/DataStore.getLists:4058/CgeoApplication.onCreate:69/Instrumentation.callApplicationOnCreate:1226/ActivityThread.handleBindApplication:7046/ActivityThread.access$1800:254}
01-14 19:47:28.794 20325 20325 I cgeo    : [main] [CtxLog]DataStore.init:END (3ms) {DataStore.initAndCheck:852/DataStore.init:812/DataStore.getLists:4058/CgeoApplication.onCreate:69/Instrumentation.callApplicationOnCreate:1226/ActivityThread.handleBindApplication:7046/ActivityThread.access$1800:254/ActivityThread$H.handleMessage:2184/Handler.dispatchMessage:106/Looper.loopOnce:233}
01-14 19:47:28.802 20325 20325 I cgeo    : [main] LooperLogger: started for Thread: main {LooperLogger.startLogging:45/CgeoApplication.onCreate:77/Instrumentation.callApplicationOnCreate:1226/ActivityThread.handleBindApplication:7046/ActivityThread.access$1800:254/ActivityThread$H.handleMessage:2184/Handler.dispatchMessage:106/Looper.loopOnce:233/Looper.loop:344/ActivityThread.main:8212}
01-14 19:47:28.802 20325 20325 I cgeo    : [main] [CtxLog]CGeoApplication.onCreate:END (15ms) {CgeoApplication.onCreate:78/Instrumentation.callApplicationOnCreate:1226/ActivityThread.handleBindApplication:7046/ActivityThread.access$1800:254/ActivityThread$H.handleMessage:2184/Handler.dispatchMessage:106/Looper.loopOnce:233/Looper.loop:344/ActivityThread.main:8212/Method.invoke:-2}
01-14 19:47:28.814 20325 20325 D cgeo    : [main] [CtxLog]SplashActivity.onCreate:START {SplashActivity.onCreate:21/Activity.performCreate:8129/Activity.performCreate:8109/Instrumentation.callActivityOnCreate:1344/ActivityThread.performLaunchActivity:3749/ActivityThread.handleLaunchActivity:3942/LaunchActivityItem.execute:109/TransactionExecutor.executeCallbacks:135/TransactionExecutor.execute:95/ActivityThread$H.handleMessage:2345}
01-14 19:47:28.818 20325 20325 V cgeo    : [main] [CtxLog]DocumentFolderAccessor.getFolderUri: /cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]:START {DocumentContentAccessor.getFolderUri:298/DocumentContentAccessor.ensureFolder:252/ContentStorage.ensureFolder:134/ContentStorage.ensureFolder:124/ContentStorage.getFolder:591/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28/Activity.performCreate:8129/Activity.performCreate:8109}
01-14 19:47:28.818 20325 20325 E ActivityThread: Failed to find provider info for com.android.externalstorage.documents
01-14 19:47:28.819 20325 20325 V cgeo    : [main] [CtxLog]DocumentFolderAccessor.getFolderUri: /cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]:END (0ms) {DocumentContentAccessor.getFolderUri:312/DocumentContentAccessor.ensureFolder:252/ContentStorage.ensureFolder:134/ContentStorage.ensureFolder:124/ContentStorage.getFolder:591/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28/Activity.performCreate:8129/Activity.performCreate:8109}
01-14 19:47:28.819 20325 20325 V cgeo    : [main] [CtxLog]DocumentFolderAccessor.getFolderUri: /cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]:START {DocumentContentAccessor.getFolderUri:298/DocumentContentAccessor.getUriForFolder:258/ContentStorage.getUriForFolder:330/LocalizationUtils.getMultiPurposeString:106/ContentStorage.reportProblem:640/ContentStorage.reportProblem:629/ContentStorage.getFolder:592/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28}
01-14 19:47:28.819 20325 20325 V cgeo    : [main] [CtxLog]DocumentFolderAccessor.getFolderUri: /cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]:END (0ms)got base(0ms); {DocumentContentAccessor.getFolderUri:312/DocumentContentAccessor.getUriForFolder:258/ContentStorage.getUriForFolder:330/LocalizationUtils.getMultiPurposeString:106/ContentStorage.reportProblem:640/ContentStorage.reportProblem:629/ContentStorage.getFolder:592/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28}
01-14 19:47:28.819 20325 20325 W cgeo    : [main] ContentStorage: Folder 'BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]](content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo)' could not be accessed {ContentStorage.reportProblem:641/ContentStorage.reportProblem:629/ContentStorage.getFolder:592/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28/Activity.performCreate:8129/Activity.performCreate:8109/Instrumentation.callActivityOnCreate:1344/ActivityThread.performLaunchActivity:3749}
01-14 19:47:28.819 20325 20325 V cgeo    : [main] [android.view.ContextThemeWrapper].showToast(Folder '/cgeo (User-defined)' could not be accessed){1} {ActivityMixin.showCgeoToast:87/ActivityMixin.showApplicationToast:111/ActivityMixin.showToast:134/ContentStorage.reportProblem:643/ContentStorage.reportProblem:629/ContentStorage.getFolder:592/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28/Activity.performCreate:8129}

**Reply by ztNFny** [reply]
just edited that log to inclode a few more lines - the "Failed to find provider info for com.android.externalstorage.documents" is again there

**Reply by ztNFny** [reply]
It seems the access check is done twice in quick succession. First one seems successful, second one fails if I read this right?

01-14 19:47:28.818 20325 20325 V cgeo    : [main] [CtxLog]DocumentFolderAccessor.getFolderUri: /cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]:START {
	DocumentContentAccessor.getFolderUri:298/
	DocumentContentAccessor.ensureFolder:252/
	ContentStorage.ensureFolder:134/
	ContentStorage.ensureFolder:124/
	ContentStorage.getFolder:591/
	ContentStorage.ensureFolder:116/
	ContentStorageActivityHelper.baseFolderIsSet:162/
	SplashActivity.onCreate:28/
	Activity.performCreate:8129/
	Activity.performCreate:8109
}
01-14 19:47:28.818 20325 20325 E ActivityThread: Failed to find provider info for com.android.externalstorage.documents
01-14 19:47:28.819 20325 20325 V cgeo    : [main] [CtxLog]DocumentFolderAccessor.getFolderUri: /cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]:START {
	DocumentContentAccessor.getFolderUri:298/
	DocumentContentAccessor.getUriForFolder:258/
	ContentStorage.getUriForFolder:330/
	LocalizationUtils.getMultiPurposeString:106/
	ContentStorage.reportProblem:640/
	ContentStorage.reportProblem:629/
	ContentStorage.getFolder:592/
	ContentStorage.ensureFolder:116/
	ContentStorageActivityHelper.baseFolderIsSet:162/
	SplashActivity.onCreate:28
}

**Reply by ztNFny** [reply]
added some debug code:
`    private Folder getFolder(final PersistableFolder folder, final boolean suppressWarningForUser) {
        Log.w("ztNFny:ContentStorage:getFolder:" + folder.getUri() + ":" + suppressWarningForUser);`

First folder that's checked is apparently "null":
01-14 20:08:12.849 13041 13041 W cgeo    : [main] ztNFny:ContentStorage:ensureFolder:null:true:true
01-14 20:08:12.917 13041 13041 W cgeo    : [main] ztNFny:ContentStorage:getFolder:content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo:false {ContentStorage.getFolder:596/ContentStorage.ensureFolder:116/ContentStorageActivityHelper.baseFolderIsSet:162/SplashActivity.onCreate:28/Activity.performCreate:8129/Activity.performCreate:8109/Instrumentation.callActivityOnCreate:1344/ActivityThread.performLaunchActivity:3749/ActivityThread.handleLaunchActivity:3942/LaunchActivityItem.execute:109}

**Reply by eddiemuc** [reply]
The second one is to create an error message for the user which should also contain the Uri as far as this is possible. I don't think this is a sign of access success.

**Reply by eddiemuc** [reply]
I am away from my laptop.

If you are willing to debug into this by adding logs: the place where the interesting stuff is happening is method DocumentContentAccessor.getFolderUri

There are diverse places inside where e.g. persisted uri permissions are checked and exceptions are silenced if thrown. Log messages should be placed there

**Reply by ztNFny** [reply]
in DocumentContentAccessor.queryDoc():
c = resolver.query(docUri, columns, null, null, null);
After installation "c == null" here, for input: docUri = "content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo" and columns = [mime_type, flags]

After granting the permissions and restarting c:geo it does "return consumer.call(c);"

As the resolver.query() goes into Android code I'm not sure how I can debug this further?

**Reply by eddiemuc** [reply]
You can't. That's pretty much the end of the line. That's super strange. As if the dir didn't exist any more

**Reply by ztNFny** [reply]
@moving-bits that's my SAF problem we discussed about in todays call. If you have any idea what else I could investigate / attempt...

Logcat shows:
ActivityThread          cgeo.geocaching                      E  Failed to find provider info for com.android.externalstorage.documents

Just before the IllegalArgumentException is thrown in DocumentContentAccessor
java.lang.IllegalArgumentException: Cursor is null when querying Uri 'content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo'
cgeo.geocaching.storage.DocumentContentAccessor.queryDoc(DocumentContentAccessor.java:458)
cgeo.geocaching.storage.DocumentContentAccessor.isValidDirectoryUri(DocumentContentAccessor.java:345)
cgeo.geocaching.storage.DocumentContentAccessor.getFolderUri(DocumentContentAccessor.java:327)
cgeo.geocaching.storage.DocumentContentAccessor.getUriForFolder(DocumentContentAccessor.java:269)
cgeo.geocaching.storage.ContentStorage.getUriForFolder(ContentStorage.java:339)
cgeo.geocaching.utils.LocalizationUtils.getMultiPurposeString(LocalizationUtils.java:106)
cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:649)
cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:638)
cgeo.geocaching.storage.ContentStorage.getFolder(ContentStorage.java:601)
cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:118)
cgeo.geocaching.storage.ContentStorageActivityHelper.baseFolderIsSet(ContentStorageActivityHelper.java:162)
cgeo.geocaching.SplashActivity.onCreate(SplashActivity.java:28)

Once SAF is re-granted that exception is no longer thrown (ie .query no longer returns null) and the "Failed to find provider info" is gone

**Reply by eddiemuc** [reply]
The IllegalArgumentException is generated by cgeo code. It is "informative" (means: handled internally and not leading to crash or other failing behavior) and thrown in exactly the case the message is explained.

The message describes the real problem: the Uri is not found by SAF (cursor is null). This is due to the content provider not found by Android system (as the first message says). And I have no clue what causes this and how to fix it. Any info would also be of high interest to me.

**Comment by eddiemuc** [comment]
Also the other log message is strange ("Failed to find provider info for com.android.externalstorage.documents") and could point to a reason of the problem.  It somehow hints that the Android content provider is not installed correctly, but in this case also the other directories should not work

**Reply by ztNFny** [reply]
Message is printed 13 times in the log, all right at start, none later when access is granted again

**Comment by eddiemuc** [comment]
A wild wild guess: could it be that this is somehow indirectly caused by the removal of the WRITE_EXTERNAL_STORAGE permission?

**Comment by ztNFny** [comment]
Just tried 2 more things: Build my build as release -> same. Use official cgeo-nightly -> same.
Am I the only one with this issue??

**Reply by moving-bits** [reply]
> Am I the only one with this issue??

During development I'm using debug builds by AS 99% of the time, and only rarely an official release candidate build. 
Debug builds can be run in the emulator (> 80%), and sometimes on my (very slow) physical testing device. Those get installed using AS "run" command.
Official release candidates are installed by downloading it and then using the system's built-in file manager to trigger a local APK install.

During all this, I get SAF prompts in fresh installs only (IIRC), mostly triggered by firing up a newly created virtual device or by switching between official release candidate and next developer build (due to different signing keys).
